tests/serac_sets.cpp

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "serac/infrastructure/input.hpp"
+#include "serac/coefficients/coefficient_extensions.hpp"
+#include <gtest/gtest.h>
+#include "mfem.hpp"
+
+#include <unordered_map>
+#include <vector>
+#include <iostream>
+#include <algorithm>
+#include "serac/infrastructure/set.hpp"
+#include "serac/physics/utilities/finite_element_state.hpp"
+#include "serac/numerics/mesh_utils.hpp"
+
+class SlicErrorException : public std::exception {
+};
+
+class SetTest : public ::testing::Test {
+protected:
+  static void SetUpTestSuite()
+  {
+    axom::slic::setAbortFunction([]() { throw SlicErrorException{}; });
+    axom::slic::setAbortOnError(true);
+    axom::slic::setAbortOnWarning(false);
+  }
+
+  void SetUp() override {}
+};
+
+namespace serac {
+
+TEST_F(SetTest, set)
+{
+  Set<int> a1(std::vector<int>{1, 2, 4, 2, 2, 3, 4});
+  std::cout << "Set a1:" << std::endl << a1 << std::endl;
+  EXPECT_EQ(a1, Set<int>({1, 2, 4, 2, 2, 3, 4}));
```

Comment:
Is this tested somewhere else to make sure you aren't getting duplicates in this case?

```
EXPECT_EQ(a1, Set<int>({1, 2, 4, 3}));
```

Reply:
I'll try to provide a better explanation of what `Set` is trying to do in the PR summary.
But essentially `Set<int>({1, 2, 4, 3})` gets converted into a std::unordered_map<attribute, index list> which would be `{ {1, {0}}, {2, {1}}, {4, {2}}, {3, {3}}`. This allows us to grab a list of elements that have a particular attribute and apply set operations to those lists.

Reply:
The `a3` example a few lines down does test for a sort of duplicate key case.

`Union` is currently defined such that in the case where there are two different sets that have an overlapping offset, the attribute the overlapping offset will be assigned to will be the attribute key that is greater. 

We could throw an error in that case, but this would require users to check using intersection/difference/complement before executing particular union calls. `Set` also tries to make it easy to create new sets from extracted subsets.

Reply:
Should this class be called `EntitySet` then to not clash/cause confusion with `std::set`?